### DISCUSSION TITLE
Connecting to a remote MCAP file distributed by nginx

### DISCUSSION THREAD

**Discussion by hedgehog319** [foxglove/community#734] [Q&A] [answered]
Hello, I am trying to access a remote **MCAP** file that I am distributing using **nginx**. When I try to connect to a remote file, I get an error:
> Error initializing: Support for HTTP Range requests was not detected on the remote file.
>
> Confirm the resource has an 'Accept-Ranges: bytes' header.
>
> Sometimes this is due to a CORS configuration error on the server. Make sure CORS is enabled with Access-Control-Allow-Origin, and that Access-Control-Expose-Headers includes Accept-Ranges.

I found [issues](https://github.com/foxglove/studio/issues/2651) with a similar problem, but it didn't help me.
Right now my nginx configuration looks like this:
```nginx
server {
        listen 80;
        server_name _;

        location / {
                alias /var/www/storage/;
                autoindex on;

                add_header Access-Control-Allow-Origin  "*";
                add_header Access-Control-Allow-Methods "GET, HEAD, OPTIONS";
                add_header Access-Control-Allow-Headers "*";
                add_header Access-Control-Expose-Headers "ETag, Content-Type, Content-Length, Accept-Ranges";
                add_header Accept-Ranges "bytes";

                if ($request_method = 'OPTIONS')
                {
                        return 204;
                }
        }
}
```
I will be grateful for your help.

**Comment by james-rms** [comment]
There is at least this guide: https://foxglove.dev/docs/studio/connection/setting-up-cors

The only difference i notice on a first pass is that "Allow-Origin" is "*" rather than https://studio.foxglove.dev. Have you tried:
```nginx
add_header Access-Control-Allow-Origin  "https://studio.foxglove.dev";
```
?

**Reply by hedgehog319** [reply]
> There is at least this guide: https://foxglove.dev/docs/studio/connection/setting-up-cors
> 
> The only difference i notice on a first pass is that "Allow-Origin" is "*" rather than https://studio.foxglove.dev. Have you tried:
> 
> ```nginx
> add_header Access-Control-Allow-Origin  "https://studio.foxglove.dev";
> ```
> 
> ?

Now a slightly different error:

> 
> Error initializing: Fetching remote file failed. TypeError: NetworkError when attempting to fetch resource.
> 
> Sometimes this is due to a CORS configuration error on the server. Make sure CORS is enabled.

**Reply by jtbandes** [reply]
Could you provide steps to reproduce the issue? Maybe using the [nginx docker image](https://hub.docker.com/_/nginx) for reproducibility?

Likely there are some more error details available in the browser's developer console but it's hard to diagnose without seeing what's there.

**Reply by hedgehog319** [reply]
Current configuration file (located at ~/nginx/default.conf):
```nginx
server {
        listen 80;
        server_name _;

        location / {
                alias /var/www/storage/;
                autoindex on;

                add_header Access-Control-Allow-Origin  "https://studio.foxglove.dev";
                add_header Access-Control-Allow-Methods "GET, HEAD, OPTIONS";
                add_header Access-Control-Allow-Headers "*";
                add_header Access-Control-Expose-Headers "ETag, Content-Type, Content-Length, Accept-Ranges";
                add_header Accept-Ranges "bytes";

                if ($request_method = 'OPTIONS')
                {
                        return 204;
                }
        }
}
```
The command to run the docker image:
`docker run --rm -d -p 80:80 -v /var/www/storage:/var/www/storage -v ~/nginx/default.conf:/etc/nginx/conf.d/default.conf --name web nginx`

There are several errors in the console:

- 404 GET https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js
- The resource from “https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js” was blocked due to MIME type (“text/plain”) mismatch (X-Content-Type-Options: nosniff).
- ChunkLoadError: Loading chunk 9554 failed. (error: https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js)

**Reply by jtbandes** [reply, marked as answer]
With one modification I got your conf file to work. You need to **remove** `add_header Accept-Ranges "bytes";`. It appears this is already set by default, and adding it again causes nginx to serve the file with `Accept-Ranges: bytes, bytes` which studio does not recognize.

If you're using HTTP(not S) you may need to allow "unsafe scripts" to get Chrome to make the request:

<img width="365" alt="image" src="https://github.com/foxglove/community/assets/14237/f38591e1-8b2a-45c1-8c4f-8060aba45f96">


> * 404 GET https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js
> * The resource from “[https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js”](https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js%E2%80%9D) was blocked due to MIME type (“text/plain”) mismatch (X-Content-Type-Options: nosniff).
> * ChunkLoadError: Loading chunk 9554 failed. (error: https://studio.foxglove.dev/9554.3601cb2219b87e69bf5d.js)

All these errors are unrelated to the nginx configuration. This would happen when the version of Studio loaded in your browser is old and a new version has been released (which can happen multiple times per day). You can fix these errors by refreshing the page.
